wallet-new/src/Cardano/Wallet/Kernel/Actions.hs

```diff
@@ -147,13 +150,52 @@ initialWorkerState = WalletWorkerState
     , _lengthPendingBlocks = 0
     }
 
--- | Start up a wallet worker; the worker will respond to actions issued over the
---   returned channel.
-forkWalletWorker :: WalletActionInterp IO b -> IO (WalletAction b -> IO ())
-forkWalletWorker ops = do
-    c <- newChan
-    link =<< async (walletWorker c ops)
-    return (writeChan c)
+
+-- | Start a wallet worker in backround who will react to input provided via the
+-- 'STM' function, in FIFO order.
+--
+-- After the given continuation returns (successfully or due to some exception),
+-- the worker will continue processing any pending input before returning,
+-- re-throwing the continuation's exception if any.
+--
+-- Usage of the obtained 'STM' action after the given continuation has returned
+-- will fail with an exception.
+withWalletWorker
+  :: (MonadIO m, Ex.MonadMask m)
+  => WalletActionInterp IO a
+  -> ((WalletAction a -> STM ()) -> m b)
+  -> m b
+withWalletWorker wai k = do
+  -- 'mDone' is full if the worker finished.
+  mDone :: MVar (Either Ex.SomeException ()) <- liftIO newEmptyMVar
+  -- 'tqWA' keeps items to be processed by the worker.
+  tqWA :: STM.TQueue (WalletAction a) <- liftIO STM.newTQueueIO
+  -- 'tvOpen' is 'True' as long as 'tqWA' can receive new input.
+  tvOpen :: STM.TVar Bool <- liftIO (STM.newTVarIO True)
+  -- 'getWA' returns the next action to be processed. This function blocks
+  -- unless 'tvOpen' is 'False', in which case 'Shutdown' is returned.
+  let getWA :: STM (WalletAction a)
+      getWA = STM.tryReadTQueue tqWA >>= \case
+         Just wa -> pure wa
+         Nothing -> STM.readTVar tvOpen >>= \case
+            False -> pure Shutdown
+            True  -> STM.retry
+  -- 'pushWA' adds an action to be executed by the worker, in FIFO order. It
+  -- will throw 'BlockedIndefinitelyOnSTM' if used after `k` returns.
+  let pushWA :: WalletAction a -> STM ()
+      pushWA = \wa -> do STM.check =<< STM.readTVar tvOpen
+                         STM.writeTQueue tqWA wa
+  Ex.mask $ \restore -> do
+     restore $ liftIO $ void $ forkFinally
+        (walletWorker wai (STM.atomically getWA))
+        (putMVar mDone)
+     Ex.finally
+        (restore (k pushWA))
+        (liftIO $ do
+            -- Prevent new input.
+            STM.atomically (STM.writeTVar tvOpen False)
+            -- Wait for the worker to finish.
+            either Ex.throwM pure =<< takeMVar mDone)
```

Comment:
I would *love* to see the practical reason around the masking here :smile: I am unsure how this differs from:

```haskell
do
  liftIO $ void $ forkFinally ...
  finally (k pushWA) (liftIO $ do ...
```

My understanding is that we mask async exceptions, and then immediately restore them for the call to `liftIO . forkFinally`. Then we call `Ex.finally` in masked state, and `restore` the action we pass. I suppose that the finalizer is also running with async exceptions masked. But, this is [already the case](https://hackage.haskell.org/package/safe-exceptions-0.1.7.0/docs/src/Control-Exception-Safe.html#finally).

Reply:
I am not sure how it differs either :smile:  I had things differently here, and changed them at last minute before submitting the PR. I think now it can be simplified as you suggest. Thanks.